### PR TITLE
drivers: bluetooth and wifi: added rs9116w

### PR DESCRIPTION
**Pull request by rtalbott-tmo**
added drivers for the rs9116w supporting both bluetooth and wifi
Signed-off-by: Rick Talbott <richard.talbott1@t-mobile.com>

This requires PR:  https://github.com/zephyrproject-rtos/hal_silabs/pull/40

### CODE REVIEW

**Review by hermabe — CHANGES_REQUESTED** [review]
It looks like you have copied the entirety of subsys/bluetooth/host into the drivers folder, this is not how to create a bluetooth driver. Please take a look at the other drivers.

File: drivers/bluetooth/CMakeLists.txt [hunk 1]
```diff
@@ -4,3 +4,4 @@ zephyr_library()
 zephyr_library_property(ALLOW_EMPTY TRUE)
 zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/bluetooth)
 add_subdirectory(hci)
+add_subdirectory(custom)
```
Comment (marcusfolkesson) [inline drivers/bluetooth/CMakeLists.txt:7]: custom is a strange name?

File: drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.c [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+		return 0;
+	}
+
+	keys = bt_keys_get_addr(id, &addr);
+	if (!keys) {
+		BT_ERR("Failed to allocate keys for %s", bt_addr_le_str(&addr));
+		return -ENOMEM;
+	}
+	if (len != BT_KEYS_STORAGE_LEN) {
+		do {
```
Comment (marcusfolkesson) [inline drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.c:288]: Please use proper error handling. do {} while(0) is not appropriate.

File: drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.c [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(bt_keys, "bt/keys", NULL, keys_set, keys_commit,
+			       NULL);
+
+#endif /* CONFIG_BT_SETTINGS */
+
+#if IS_ENABLED(CONFIG_BT_KEYS_OVERWRITE_OLDEST)
+void bt_keys_update_usage(uint8_t id, const bt_addr_le_t *addr)
```
Comment (marcusfolkesson) [inline drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.c:347]: Is this function used at all?

File: drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.h [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+struct bt_keys_link_key *bt_keys_find_link_key(const bt_addr_t *addr);
+void bt_keys_link_key_clear(struct bt_keys_link_key *link_key);
+void bt_keys_link_key_clear_addr(const bt_addr_t *addr);
+void bt_keys_link_key_store(struct bt_keys_link_key *link_key);
+
+
+/* This function is used to signal that the key has been used for paring */
+/* It updates the aging counter and saves it to flash if configuration option */
+/* BT_KEYS_SAVE_AGING_COUNTER_ON_PAIRING is enabled */
+void bt_keys_update_usage(uint8_t id, const bt_addr_le_t *addr);
```
Comment (marcusfolkesson) [inline drivers/bluetooth/custom/rs9116w/rs9116w_ble_keys.h:125]: The definition of bt_keys_update_usage() is encapsulated in within #if (defined(CONFIG_BT_KEYS_OVERWRITE_OLDEST))-block. So should the prototype?

File: drivers/bluetooth/custom/rs9116w/rs9116w_ble_shell.c [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+static int ble_adv_stop(const struct shell *shell, int argc, char **argv)
+{
+	if (!RS9116_EN) {
+		shell_error(shell, "Device not enabled");
+		return 1;
+	}
+	int err;
+
+	err = bt_le_adv_stop();
+	if (err) {
```
Comment (marcusfolkesson) [inline drivers/bluetooth/custom/rs9116w/rs9116w_ble_shell.c:53]: if (bt_le_adv_stop()) ?

**Review by marcusfolkesson — COMMENTED** [review]
An general comment, there are several places like

err = foo();
if (err)
    return 1;
else
   return 0;

The err usually contains an error code, it is better to propagate those errors instead of returning 1.

**Comment by rtalbott-tmo** [comment]
> It looks like you have copied the entirety of subsys/bluetooth/host into the drivers folder, this is not how to create a bluetooth driver. Please take a look at the other drivers.

It's unfortunate that that chip does not provide a HCI or L2CAP interface.   So it was necessary to re-use a lot of existing host functions to interface with.

File: drivers/bluetooth/CMakeLists.txt (hunk 1 above)
Comment (jtbaumann) [reply drivers/bluetooth/CMakeLists.txt:7]: The reason for the name custom is that this driver is not HCI based, unlike the existing bluetooth drivers within zephyr. There is existing entries in the configs for custom drivers, so we simply added a folder for custom drivers and located our driver within that folder.

**Comment by rtalbott-tmo** [comment]
Will break this apart into two different PRs.  Fist will be wifi, next will be bluetooth.

**Pull request closed without merging**